File: OTCDerivativesCalculatorModule/Project_CSharp/XmlBuilder/XmlSerializedClass/fpml-cd-5-4/ValuationDate.cs
using System;
using System.Text;
using System.Collections.Generic;
using System.Xml;

namespace FpmlSerializedCSharp
{
    public class ValuationDate : ISerialized
    {
        public ValuationDate(XmlNode xmlNode)
        {
            XmlNode singleValuationDateNode = xmlNode.SelectSingleNode("singleValuationDate");
            
            if (singleValuationDateNode != null)
            {
                if (singleValuationDateNode.Attributes["href"] != null || singleValuationDateNode.Attributes["id"] != null) 
                {
                    if (singleValuationDateNode.Attributes["id"] != null) 
                    {
                        singleValuationDateIDRef_ = singleValuationDateNode.Attributes["id"].Value;
                        SingleValuationDate ob = new SingleValuationDate(singleValuationDateNode);
                        IDManager.SetID(singleValuationDateIDRef_, ob);
                    }
                    else if (singleValuationDateNode.Attributes["href"] != null)
                    {
                        singleValuationDateIDRef_ = singleValuationDateNode.Attributes["href"].Value;
                    }
                    else
                    {
                        singleValuationDate_ = new SingleValuationDate(singleValuationDateNode);
                    }
                }
                else
                {
                    singleValuationDate_ = new SingleValuationDate(singleValuationDateNode);
                }
            }
            
        
            XmlNode multipleValuationDatesNode = xmlNode.SelectSingleNode("multipleValuationDates");
            
            if (multipleValuationDatesNode != null)
            {
                if (multipleValuationDatesNode.Attributes["href"] != null || multipleValuationDatesNode.Attributes["id"] != null) 
                {
                    if (multipleValuationDatesNode.Attributes["id"] != null) 
                    {
                        multipleValuationDatesIDRef_ = multipleValuationDatesNode.Attributes["id"].Value;
                        MultipleValuationDates ob = new MultipleValuationDates(multipleValuationDatesNode);
                        IDManager.SetID(multipleValuationDatesIDRef_, ob);
                    }
                    else if (multipleValuationDatesNode.Attributes["href"] != null)
                    {
                        multipleValuationDatesIDRef_ = multipleValuationDatesNode.Attributes["href"].Value;
                    }
                    else
                    {
                        multipleValuationDates_ = new MultipleValuationDates(multipleValuationDatesNode);
                    }
                }
                else
                {
                    multipleValuationDates_ = new MultipleValuationDates(multipleValuationDatesNode);
                }
            }
            
        
        }
        
    
        #region SingleValuationDate_
        private SingleValuationDate singleValuationDate_;
        public SingleValuationDate SingleValuationDate_
        {
            get
            {
                if (this.singleValuationDate_ != null)
                {
                    return this.singleValuationDate_; 
                }
                else if (this.singleValuationDateIDRef_ != null)
                {
                    singleValuationDate_ = IDManager.getID(singleValuationDateIDRef_) as SingleValuationDate;
                    return this.singleValuationDate_; 
                }
                else
                {
                      return this.singleValuationDate_; 
                }
            }
            set
            {
                if (this.singleValuationDate_ != value)
                {
                    this.singleValuationDate_ = value;
                }
            }
        }
        #endregion
        
        public string singleValuationDateIDRef_ { get; set; }
        #region MultipleValuationDates_
        private MultipleValuationDates multipleValuationDates_;
        public MultipleValuationDates MultipleValuationDates_
        {
            get
            {
                if (this.multipleValuationDates_ != null)
                {
                    return this.multipleValuationDates_; 
                }
                else if (this.multipleValuationDatesIDRef_ != null)
                {
                    multipleValuationDates_ = IDManager.getID(multipleValuationDatesIDRef_) as MultipleValuationDates;
                    return this.multipleValuationDates_; 
                }
                else
                {
                      return this.multipleValuationDates_; 
                }
            }
            set
            {
                if (this.multipleValuationDates_ != value)
                {
                    this.multipleValuationDates_ = value;
                }
            }
        }
        #endregion
        
        public string multipleValuationDatesIDRef_ { get; set; }
        public string choiceStr_0;
        
    
        
    
    }
    
}

File: OTCDerivativesCalculatorModule/Project_Cpp/lib_static/FpmlSerialized/GenClass/fpml-cd-5-4/ValuationDate.cpp
// ValuationDate.cpp 
#include "ValuationDate.hpp"
#ifdef ConsolePrint
    #include <iostream>
#endif 

namespace FpmlSerialized {

ValuationDate::ValuationDate(TiXmlNode* xmlNode)
: ISerialized(xmlNode)
{
    #ifdef ConsolePrint
        std::string initialtap_ = FileManager::instance().tap_;
        FileManager::instance().tap_.append("   ");
    #endif 
   //singleValuationDateNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* singleValuationDateNode = xmlNode->FirstChildElement("singleValuationDate");

   if(singleValuationDateNode){singleValuationDateIsNull_ = false;}
   else{singleValuationDateIsNull_ = true;}

   #ifdef ConsolePrint
      FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- singleValuationDateNode , address : " << singleValuationDateNode << std::endl;
   #endif
   if(singleValuationDateNode)
   {
      if(singleValuationDateNode->Attribute("href") || singleValuationDateNode->Attribute("id"))
      {
          if(singleValuationDateNode->Attribute("id"))
          {
             singleValuationDateIDRef_ = singleValuationDateNode->Attribute("id");
             singleValuationDate_ = boost::shared_ptr<SingleValuationDate>(new SingleValuationDate(singleValuationDateNode));
             singleValuationDate_->setID(singleValuationDateIDRef_);
             IDManager::instance().setID(singleValuationDateIDRef_,singleValuationDate_);
          }
          else if(singleValuationDateNode->Attribute("href")) { singleValuationDateIDRef_ = singleValuationDateNode->Attribute("href");}
          else { QL_FAIL("id or href error"); }
      }
      else { singleValuationDate_ = boost::shared_ptr<SingleValuationDate>(new SingleValuationDate(singleValuationDateNode));}
   }

   //multipleValuationDatesNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* multipleValuationDatesNode = xmlNode->FirstChildElement("multipleValuationDates");

   if(multipleValuationDatesNode){multipleValuationDatesIsNull_ = false;}
   else{multipleValuationDatesIsNull_ = true;}

   #ifdef ConsolePrint
      FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- multipleValuationDatesNode , address : " << multipleValuationDatesNode << std::endl;
   #endif
   if(multipleValuationDatesNode)
   {
      if(multipleValuationDatesNode->Attribute("href") || multipleValuationDatesNode->Attribute("id"))
      {
          if(multipleValuationDatesNode->Attribute("id"))
          {
             multipleValuationDatesIDRef_ = multipleValuationDatesNode->Attribute("id");
             multipleValuationDates_ = boost::shared_ptr<MultipleValuationDates>(new MultipleValuationDates(multipleValuationDatesNode));
             multipleValuationDates_->setID(multipleValuationDatesIDRef_);
             IDManager::instance().setID(multipleValuationDatesIDRef_,multipleValuationDates_);
          }
          else if(multipleValuationDatesNode->Attribute("href")) { multipleValuationDatesIDRef_ = multipleValuationDatesNode->Attribute("href");}
          else { QL_FAIL("id or href error"); }
      }
      else { multipleValuationDates_ = boost::shared_ptr<MultipleValuationDates>(new MultipleValuationDates(multipleValuationDatesNode));}
   }

    #ifdef ConsolePrint
        FileManager::instance().tap_ = initialtap_;
    #endif 
}
boost::shared_ptr<SingleValuationDate> ValuationDate::getSingleValuationDate()
{
   if(!this->singleValuationDateIsNull_){
        if(singleValuationDateIDRef_ != ""){
             return boost::shared_static_cast<SingleValuationDate>(IDManager::instance().getID(singleValuationDateIDRef_));
        }else{
             return this->singleValuationDate_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return boost::shared_ptr<SingleValuationDate>();
   }
}
boost::shared_ptr<MultipleValuationDates> ValuationDate::getMultipleValuationDates()
{
   if(!this->multipleValuationDatesIsNull_){
        if(multipleValuationDatesIDRef_ != ""){
             return boost::shared_static_cast<MultipleValuationDates>(IDManager::instance().getID(multipleValuationDatesIDRef_));
        }else{
             return this->multipleValuationDates_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return boost::shared_ptr<MultipleValuationDates>();
   }
}
}

